src/Symfony/Component/Message/DependencyInjection/MessagePass.php

```diff
@@ -0,0 +1,139 @@
+{
+    use PriorityTaggedServiceTrait;
+
+    private $messageBusService;
+    private $messageHandlerResolverService;
+    private $handlerTag;
+
+    public function __construct(string $messageBusService = 'message_bus', string $messageHandlerResolverService = 'message.handler_resolver', string $handlerTag = 'message_handler')
+    {
+        $this->messageBusService = $messageBusService;
+        $this->messageHandlerResolverService = $messageHandlerResolverService;
+        $this->handlerTag = $handlerTag;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->hasDefinition($this->messageBusService)) {
+            return;
+        }
+
+        if (!$container->getParameter('kernel.debug') || !$container->hasDefinition('logger')) {
+            $container->removeDefinition('message.middleware.debug.logging');
+        }
+
+        $this->registerReceivers($container);
+        $this->registerHandlers($container);
+    }
+
+    private function registerHandlers(ContainerBuilder $container)
+    {
+        $handlersByMessage = array();
+
+        foreach ($container->findTaggedServiceIds($this->handlerTag, true) as $serviceId => $tags) {
+            foreach ($tags as $tag) {
+                $handles = $tag['handles'] ?? $this->guessHandledClass($container, $serviceId);
+
+                if (!class_exists($handles)) {
+                    $messageClassLocation = isset($tag['handles']) ? 'declared in your tag attribute "handles"' : 'declared in `__invoke` function';
+
+                    throw new RuntimeException(sprintf('The message class "%s" %s of service "%s" does not exist.', $messageClassLocation, $handles, $serviceId));
+                }
+
+                $priority = $tag['priority'] ?? 0;
+                $handlersByMessage[$handles][$priority][] = new Reference($serviceId);
+            }
+        }
+
+        foreach ($handlersByMessage as $message => $handlers) {
+            krsort($handlersByMessage[$message]);
+            $handlersByMessage[$message] = call_user_func_array('array_merge', $handlersByMessage[$message]);
+        }
+
+        $definitions = array();
+        foreach ($handlersByMessage as $message => $handlers) {
+            if (1 === count($handlers)) {
+                $handlersByMessage[$message] = current($handlers);
+            } else {
+                $d = new Definition(ChainHandler::class, array($handlers));
+                $d->setPrivate(true);
+                $serviceId = hash('sha1', $message);
+                $definitions[$serviceId] = $d;
+                $handlersByMessage[$message] = new Reference($serviceId);
+            }
+        }
+        $container->addDefinitions($definitions);
+
+        $handlersLocatorMapping = array();
+        foreach ($handlersByMessage as $message => $handler) {
+            $handlersLocatorMapping['handles.'.$message] = $handler;
```

Comment:
Isn't this one supposed to be `handler` just like in ContainerHandlerLocator.php line 33?